drivers/i2c/i2c_esp32.c

```diff
@@ -41,6 +43,14 @@ LOG_MODULE_REGISTER(i2c_esp32, CONFIG_I2C_LOG_LEVEL);
 #define I2C_CLK_LIMIT_RTC (20 * 1000 * 1000 / 20)	/* Limited by RTC, no more than RTC/20 */
 #define I2C_CLK_LIMIT_XTAL (40 * 1000 * 1000 / 20)	/* Limited by RTC, no more than XTAL/20 */
 
+#define I2C_GET_PIN_INFO(state, idx)								\
+{												\
+	.gpio_name = gpio_esp32_get_gpio_for_pin(ESP32_PIN_NUM(state->pins[idx].pinmux)),	\
+	.sig_out = ESP32_PIN_SIGO(state->pins[idx].pinmux),					\
+	.sig_in = ESP32_PIN_SIGI(state->pins[idx].pinmux),					\
+	.pin = ESP32_PIN_NUM(state->pins[idx].pinmux)						\
+}
```

Comment:
it is not allowed to query pin information, API operates on states. all pinctrl structs should be considered private.

Reply:
This query is not part of the regular flow of the pinctrl API. It is included here to overcome an I2C driver constraint:

Most Espressif SoCs are provisioned with HAL mechanisms to overcome an eventual I2C FSM failure, while others need to keep pin numbering info to achieve the same.

Reply:
The problem here is that driver's can't make use of any pinctrl state information since it is opaque. You can't rely on the data structure behind pinctrl_dev_config. If you need GPIO access to i2c pins, you'll need something like:

1. Create a new custom state "gpio" (or whatever name) that configures pins as GPIO instead of muxed to the I2C peripheral
2. Store GPIO phandles in DT, e.g. `i2c-gpios = <&gpiox 0 1>, <&gpioy 2 3>;`. It's true that this duplicates some information also stored by pinctrl, but right now, the pinctrl API doesn't expose any specified pin as a GPIO device. We may explore this in the future based on the initial round of pinctrl ports.
3. When in need for gpio access, just do `pinctrl_apply_state(..., PINCTRL_GPIO_STATE);`. Use GPIO API. When done, apply `PINCTRL_DEFAULT_STATE` back.

Reply:
I have gone through a different approach but got rid of the pin state query.